src/clippoly/src/clippoly.cpp

```diff
@@ -0,0 +1,271 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "clippoly.h"
+
+#include <clipper2/clipper.h>
+
+#include <vector>
+
+#include "clipper2/clipper.core.h"
+#include "clipper2/clipper.engine.h"
+#include "glm/ext/vector_float2.hpp"
+#include "public.h"
+
+using namespace manifold;
+
+namespace {
+const int precision_ = 8;
+
+C2::ClipType cliptype_of_op(Clippoly::OpType op) {
+  C2::ClipType ct = C2::ClipType::Union;
+  switch (op) {
+    case Clippoly::OpType::Add:
+      break;
+    case Clippoly::OpType::Subtract:
+      ct = C2::ClipType::Difference;
+      break;
+    case Clippoly::OpType::Intersect:
+      ct = C2::ClipType::Intersection;
+      break;
+    case Clippoly::OpType::Xor:
+      ct = C2::ClipType::Xor;
+      break;
+  };
+  return ct;
+}
+C2::PathD pathd_of_contour(std::vector<glm::vec2> ctr) {
+  auto p = C2::PathD();
+  for (auto v : ctr) {
+    p.push_back(C2::PointD(v.x, v.y));
+  }
+  return p;
+}
+}  // namespace
+
+namespace manifold {
+Clippoly::Clippoly() { paths_ = C2::PathsD(); }
+Clippoly::~Clippoly() = default;
+Clippoly::Clippoly(Clippoly&&) noexcept = default;
+Clippoly& Clippoly::operator=(Clippoly&&) noexcept = default;
+Clippoly::Clippoly(const Clippoly& other) {
+  paths_ = C2::PathsD(other.paths_);
+  clean_ = other.clean_;
+}
+Clippoly::Clippoly(C2::PathsD ps, bool clean) {
+  paths_ = ps;
+  clean_ = clean;
+}
+
+Clippoly::Clippoly(std::vector<glm::vec2> contour) {
+  auto ps = C2::PathsD();
+  ps.push_back(pathd_of_contour(contour));
+  paths_ = ps;
+}
+
+Clippoly::Clippoly(std::vector<std::vector<glm::vec2>> contours) {
+  auto ps = C2::PathsD();
+  for (auto ctr : contours) {
+    ps.push_back(pathd_of_contour(ctr));
+  }
+  paths_ = ps;
+}
+
+Clippoly Clippoly::Square(glm::vec2 dims, bool center) {
+  auto p = C2::PathD();
+  if (center) {
+    auto w = dims.x / 2;
+    auto h = dims.y / 2;
+    p.push_back(C2::PointD(w, -h));
+    p.push_back(C2::PointD(-w, -h));
+    p.push_back(C2::PointD(-w, h));
+    p.push_back(C2::PointD(w, h));
+  } else {
+    double x = dims.x;
+    double y = dims.y;
+    p.push_back(C2::PointD(0.0, y));
+    p.push_back(C2::PointD(x, y));
+    p.push_back(C2::PointD(x, 0.0));
+    p.push_back(C2::PointD(0.0, 0.0));
+  }
+  auto ps = C2::PathsD();
+  ps.push_back(p);
+  return Clippoly(ps, true);
+}
+
+Clippoly Clippoly::Circle(float radius, int circularSegments) {
+  // GetCircularSegments(radius) -- in Manifold, not available atm
+  int n = circularSegments > 2 ? circularSegments : 3;
+  float dPhi = 360.0f / n;
+  auto circle = C2::PathsD(1);
+  for (int i = 0; i < n; ++i) {
+    circle[0].push_back(
+        C2::PointD(radius * cosd(dPhi * i), radius * sind(dPhi * i)));
+  }
+  return Clippoly(circle, true);
+}
+
+Clippoly Clippoly::Boolean(const Clippoly& second, OpType op) const {
+  auto ct = cliptype_of_op(op);
+  auto res = C2::BooleanOp(ct, C2::FillRule::NonZero, paths_, second.paths_,
```

Comment:
Does clipper have a strictly-positive fill rule? I generally prefer that to non-zero.

Reply:
It does, but OpenSCAD uses non-zero, so I have been favouring that.

Reply:
We don't follow *everything* OpenSCAD does - note our sphere is quite a bit different (and better) than theirs. In the event of self-intersection, non-zero gives xor-like polygons that share verts. It also discards winding direction information, since it's symmetric. The 3MF spec uses strictly-positive for this reason, so I think we should too.

Reply:
If push comes to shove, we can always add an optional construction parameter to specify the desired fill rule.